### DISCUSSION TITLE
为什么 MD5 不能用于存储密码 · Why's THE Design? · /whys-the-design-password-with-md5

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#575] [Announcements]
https://draveness.me/whys-the-design-password-with-md5

**Comment by leafduo** [comment]
我印象中一些新的散列算法还有一些不易在 GPU 中实现、不易硬件实现之类的特性

**Comment by draveness** [comment]
> 我印象中一些新的散列算法还有一些不易在 GPU 中实现、不易硬件实现之类的特性

确实有，区块链里很多算法就是设计抗 GPU 的，也能搜索到一些相关的论文 [RandomHash:
GPU & ASIC Resistant Hash Algorithm](https://www.pascalcoin.org/storage/whitepapers/RandomHash_Whitepaper.pdf)，不过没细研究过，不是特别了解

**Comment by gitwillsky** [comment]
德莱文为何如此🐂，什么都懂

**Comment by qqqq331792873** [comment]
我读了两遍"哈希加盐",可是还是没读懂为什么不能用哈希加盐的形式存储密码.

**Comment by draveness** [comment]
> 我读了两遍"哈希加盐",可是还是没读懂为什么不能用哈希加盐的形式存储密码.

哈希加盐的密码会被目前的硬件迅速破解

**Comment by z-wade** [comment]
`MD5`容易被彩虹表破解，优化成`加盐MD5`。
`加盐MD5`的在`GPU`比较牛逼的情况下，哈希碰撞容易破解。其实只用`MD5`也会通过哈希碰撞破解，至于为啥不用，应该是彩虹表更加高效。
最后`bcrypt`也是不可逆的哈希算法，但是它有个特点，当计算大量密码的时候速度比较慢，大大提高了攻击成本。
但是我有个问题：`bcrypt`也算是哈希算法，所以也会被彩虹表破解，是不是得加盐才更加安全？

**Comment by draveness** [comment]
> 但是我有个问题：`bcrypt`也算是哈希算法，所以也会被彩虹表破解，是不是得加盐才更加安全？

bcrypt 有内置的盐来避免彩虹表的攻击

> bcrypt has salts built-in to prevent rainbow table attacks.

+ http://codahale.com/how-to-safely-store-a-password/
+ https://stackoverflow.com/questions/6832445/how-can-bcrypt-have-built-in-salts

**Comment by daoshenzzg** [comment]
> > 我读了两遍"哈希加盐",可是还是没读懂为什么不能用哈希加盐的形式存储密码.
> 
> 哈希加盐的密码会被目前的硬件迅速破解

hash(salt, value) = hash(salt, password)』、『他只需要找到一个值 value』

他应该还需要找到合适的 salt 吧？

**Comment by draveness** [comment]
> > > 我读了两遍"哈希加盐",可是还是没读懂为什么不能用哈希加盐的形式存储密码.
> > 
> > 
> > 哈希加盐的密码会被目前的硬件迅速破解
> 
> hash(salt, value) = hash(salt, password)』、『他只需要找到一个值 value』
> 
> 他应该还需要找到合适的 salt 吧？

salt 在密码里存的都是明文

secret = hash(salt, password) + salt

**Comment by wenzhiquan** [comment]
如果是hash(salt + hash(password))的方式，或者hash(salt + hash(password)[0:8])这种方式，让攻击者无法知道你的hash值计算方式，是否就更加安全了呢？

**Comment by draveness** [comment]
> 如果是hash(salt + hash(password))的方式，或者hash(salt + hash(password)[0:8])这种方式，让攻击者无法知道你的hash值计算方式，是否就更加安全了呢？

把 salt 也哈希了，那怎么验证密码呢...

**Comment by wenzhiquan** [comment]
> @draveness
> > 如果是hash(salt + hash(password))的方式，或者hash(salt + hash(password)[0:8])这种方式，让攻击者无法知道你的hash值计算方式，是否就更加安全了呢？
> 
> 把 salt 也哈希了，那怎么验证密码呢...

数据库的password存的是hash(salt + hash(password))后的值，salt是明文存储，收到password之后使用hash(salt + hash(password))计算出的值和数据库存的password进行比对，如果一致就是正确的密码。这样攻击者并不知道你是如何hash出的值，彩虹表和hash碰撞也就很难破解这个密码了吧

**Comment by draveness** [comment]
> > @draveness
> > > 如果是hash(salt + hash(password))的方式，或者hash(salt + hash(password)[0:8])这种方式，让攻击者无法知道你的hash值计算方式，是否就更加安全了呢？
> > 
> > 
> > 把 salt 也哈希了，那怎么验证密码呢...
> 
> 数据库的password存的是hash(salt + hash(password))后的值，salt是明文存储，收到password之后使用hash(salt + hash(password))计算出的值和数据库存的password进行比对，如果一致就是正确的密码。这样攻击者并不知道你是如何hash出的值，彩虹表和hash碰撞也就很难破解这个密码了吧

这相当于发明一种『新的』加密策略，现在的硬件计算能力非常恐怖的，别人知道你的策略后破译起来非常简单，哈希两次与哈希一次没有太多的区别。

密码学上有句话：『永远不要自己发明加密算法』，只有经过业界长时间验证的算法才是安全的，还是直接使用 bcrypt 吧。

**Comment by jadynwu** [comment]
有个网站叫CMD5，里面存了很多密码和MD5值的映射，用来反查，服务器稳定运行了十几年

另外测了一下sha1的性能更优，因为sha256的运算时间是sha1的2倍，如果考虑性能可以用sha1

**Comment by draveness** [comment]
> 有个网站叫CMD5，里面存了很多密码和MD5值的映射，用来反查，服务器稳定运行了十几年

感谢分享

> 另外测了一下sha1的性能更优，因为sha256的运算时间是sha1的2倍，如果考虑性能可以用sha1

加密还是应该改用 bcrypt，但是哈希的话要看场景，安全重要还是性能重要。不过多数场景下稍微慢一些不会有太大影响的

**Comment by jadynwu** [comment]
hicker攻击数据库，只要拿到账号和密码（加密、hash、bcrype等任何一种形式复杂过的值）的映射，通过登录校验就行了，作者这篇文章讨论的安全，真正的作用是为了保护用户的明文密码的隐私不被泄漏，比如hicker用来登录用户的其他平台账号对吗？  hicker如果只是为了获取到大量使用户登录校验通过的信息，这种只对密码进行bcript的方式似乎没有用。

**Comment by jadynwu** [comment]
> @qqqq331792873
> 我读了两遍"哈希加盐",可是还是没读懂为什么不能用哈希加盐的形式存储密码.

个人觉得其实可以，MD5反向其实只能通过查已知表做到，这个已知表是服务器跑出来的，盐只要足够复杂，几乎不可能反向查出密码明文

**Comment by draveness** [comment]
> hicker攻击数据库，只要拿到账号和密码（加密、hash、bcrype等任何一种形式复杂过的值）的映射，通过登录校验就行了，作者这篇文章讨论的安全，真正的作用是为了保护用户的明文密码的隐私不被泄漏，比如hicker用来登录用户的其他平台账号对吗？ hicker如果只是为了获取到大量使用户登录校验通过的信息，这种只对密码进行bcript的方式似乎没有用。

如果攻击者只拿到 bcrypt 『哈希』后的密码，它能在较短的时间内找到原文或者与原文加密结果相同的字符串？这篇文章讨论的也只是怎么加密密码更安全，如果说通过别的方式伪装用户登录（例如：拿到用户的 AuthToken），那与这篇文章讨论的内容就没有太多关系了。

> 个人觉得其实可以，MD5反向其实只能通过查已知表做到，这个已知表是服务器跑出来的，盐只要足够复杂，几乎不可能反向查出密码明文

是不能反查出来，因为盐是一般是与密码一起存储的，所以攻击者拿到了加密后的值可以使用搞算力的硬件迅速破解。

**Comment by jadynwu** [comment]
> @draveness
> > hicker攻击数据库，只要拿到账号和密码（加密、hash、bcrype等任何一种形式复杂过的值）的映射，通过登录校验就行了，作者这篇文章讨论的安全，真正的作用是为了保护用户的明文密码的隐私不被泄漏，比如hicker用来登录用户的其他平台账号对吗？ hicker如果只是为了获取到大量使用户登录校验通过的信息，这种只对密码进行bcript的方式似乎没有用。
> 
> 如果攻击者只拿到 bcrypt 『哈希』后的密码，它能在较短的时间内找到原文或者与原文加密结果相同的字符串？这篇文章讨论的也只是怎么加密密码更安全，如果说通过别的方式伪装用户登录（例如：拿到用户的 AuthToken），那与这篇文章讨论的内容就没有太多关系了。
> 
> > 个人觉得其实可以，MD5反向其实只能通过查已知表做到，这个已知表是服务器跑出来的，盐只要足够复杂，几乎不可能反向查出密码明文
> 
> 是不能反查出来，因为盐是一般是与密码一起存储的，所以攻击者拿到了加密后的值可以使用搞算力的硬件迅速破解。

首先
对于你说的模拟登陆不在本文范畴这点我理解了，但是我觉得，你本篇文章对攻击者的目的描述有误，攻击者的目的不应该是拿到加密结果相同的字符串，而是要拿到密码本身，因为密码明文对攻击者的作用似乎就是想要用在其他的地方，因为攻击者能直接在你的数据库拿到加密或hash后的结果，不需要知道加密过程就可以在你的服务商进行模拟登陆（账号 + 密码密文），由于不能逆向，你是拿密文验证登陆密码是否是正确的。

第二
“搞算力的硬件迅速破解”这点不太明白，除了MD5以外，其他的hash方式没有碰撞，CMD5这种网站其实就是属于用高算力的硬件迅速遍历的破解，但是被破解的都是简单密码，除非存储在服务器的盐被看到了，攻击者通过加同样的盐进行破解。即便是这样，也需要在简单密码的前提下，才有可能被暴力破解

**Comment by draveness** [comment]
> 你本篇文章对攻击者的目的描述有误，攻击者的目的不应该是拿到加密结果相同的字符串，而是要拿到密码本身，因为密码明文对攻击者的作用似乎就是想要用在其他的地方

『似乎是』就是不是，拿到明文可以用在所有网站，但是拿到哈希结果相同的其他原文就可以在当前网站登录。

> “搞算力的硬件迅速破解”这点不太明白，除了MD5以外，其他的hash方式没有碰撞，CMD5这种网站其实就是属于用高算力的硬件迅速遍历的破解，但是被破解的都是简单密码，除非存储在服务器的盐被看到了，攻击者通过加同样的盐进行破解。即便是这样，也需要在简单密码的前提下，才有可能被暴力破解

可以了解下 hashcat

**Comment by jackiect** [comment]
对于不同的用户使用不同的salt呢？比如用户名，因为业务中用户名通常是不允许变更的

**Comment by draveness** [comment]
> 对于不同的用户使用不同的salt呢？比如用户名，因为业务中用户名通常是不允许变更的

直接用标准就行了，别的方法都没经过广泛验证，肯定没有标准的存储策略安全..

**Comment by 000b000** [comment]
受教了，多谢博主

**Comment by PeterlitsZo** [comment]
我还是感觉好奇怪呀~

如果说加盐的话，用户密码太普通还是会被常用密码表给搞下来呀？因为盐是明文，给定一个密码password，和数据库储存的加盐摘要hash(salt + password)，那么我用哈希表判断

```python
[hash(salt + password) = hashed_data for password in table]
```

也是可以判断出来的呀，毕竟验证是否是正确密码本身就是一个纯函数，而知道输出和输入分布，我按照频率从大到小还是能找出很多用户的密码呀？

**Comment by hoperuin** [comment]
https://md5hashing.net/hash/md5/23cdc18507b52418db7740cbb5543e54

**Comment by hyzgh** [comment]
假如用sha256来存储密码，即使可以通过 GPU 每秒执行上亿次的计算，但是也需要2^256/100000000/60/60/24/365=3.67e61年才能破解，没太理解为什么不能用哈希函数来存储密码

**Comment by hyzgh** [comment]
使用哈希加盐存储密码。有研究标明发现了MD5的快速碰撞方法，所以不安全。但是我没理解SHA256为什么不可以。

**Comment by lecssmi** [comment]
> @hyzgh
> 假如用sha256来存储密码，即使可以通过 GPU 每秒执行上亿次的计算，但是也需要2^256/100000000/60/60/24/365=3.67e61年才能破解，没太理解为什么不能用哈希函数来存储密码

应该是理论上的不安全吧。毕竟大部分公司都没有到达那种安全级别。
